### ISSUE TITLE
CSI PVC restoring as-is with same original volume handle when VolumeSnapshot is not present with RestorePVs as true

### ISSUE THREAD

**Issue by mayankagg9722** [vmware-tanzu/velero#8493]
**What steps did you take and what happened:**
1. Having CSI dynamic PVC and PVs in the cluster. Performing velero backup with  _[snapshotVolumes](https://github.com/vmware-tanzu/velero/blob/aa7ca1515926fa029e4d2f4d12e712f5e639a9ef/pkg/cmd/cli/backup/create.go#L84): false_
2. Restoring with this backup in the same cluster to different namespace. (performing namespace mapping). 

**Behaviour during Backup:** Volume Snapshot are not taken as snapshotVolumes is set to False, but PVC YAMLs are backed up.

**Issue During Restore:** During restore velero is also restoring PVC Yamls as-is in the cluster with the same original volume handle, causing conflict if original PVC already exists in the cluster when [RestorePVs](https://github.com/vmware-tanzu/velero/blob/aa7ca1515926fa029e4d2f4d12e712f5e639a9ef/pkg/apis/velero/v1/restore_types.go#L89)=true 

![image](https://github.com/user-attachments/assets/74a39a06-0197-4c4d-a4ab-b617f1ff592a)

There are 2 different behaviours currently with velero restore:

1. RestorePVs=false

Here, restore of PVC yaml happens but velero is removing the volumehandle which is leading to new PV getting provisioned in the cluster ( this is the right behaviour).

https://github.com/vmware-tanzu/velero/blob/aa7ca1515926fa029e4d2f4d12e712f5e639a9ef/pkg/restore/actions/csi/pvc_action.go#L178

2. RestorePVs=true

Here, restore of PVC yaml happening with same original volume handle in the cluster and velero is not removing the volumehandle which is leading to PVC provisioning failed (we need to improve this behaviour).

https://github.com/vmware-tanzu/velero/blob/aa7ca1515926fa029e4d2f4d12e712f5e639a9ef/pkg/restore/actions/csi/pvc_action.go#L224

Creation of PVC Yaml during restore:
https://github.com/vmware-tanzu/velero/blob/aa7ca1515926fa029e4d2f4d12e712f5e639a9ef/pkg/restore/restore.go#L1513C3-L1513C33

**What did you expect to happen:**
We expect volume handle from the PVC should be removed so that dynamic volume creation can happen at the time of restore of PVC YAML even when restored with flag RestorePVs=true when volume snapshot is not present.

Error:

![image](https://github.com/user-attachments/assets/6bc89097-a74c-42f9-998a-f4a0492a12a0)

**Logs:**

**RestorePVs=false**
![image](https://github.com/user-attachments/assets/67559864-bbdc-4bda-beaf-270edb8d2aa6)

**RestorePVs=true**

![image](https://github.com/user-attachments/assets/65371fb8-e343-4466-94f3-0ad30ff91405)

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by ywk253100** [comment]
@mayankagg9722 Could you explain a bit more about why you expect `RestorePVs=true` to work?
Per my understanding, `RestorePVs=false` already meets your requirement and `RestorePVs=ture` works as design.

**Comment by kaovilai** [comment]
> We expect volume handle from the PVC should be removed so that dynamic volume creation can happen at the time of restore of PVC YAML even when restored with flag RestorePVs=true when volume snapshot is not present.

If you completely delete NS containing PVC and PV and restore from backup.. I would expect the same PV name to be created and used when not restoring from a snapshot.

The example scenario only happens if your backup inclulde PVC and PV yaml and you restore to a different NS perhaps right?

**Comment by mayankagg9722** [comment]
Yes @kaovilai you are right. So, is this the expected behaviour?

**Comment by kaovilai** [comment]
As currently implemented yes. Velero does not have pv name conflict resolution that I'm aware of.

Only scenarios that do not involve defining pv name such as creating PVC from dataSource snapshot.

**Comment by mayankagg9722** [comment]
@kaovilai  I think we can make a similar change to make volume handling nil explicitly during this scenario as well when there is no snapshot exists: https://github.com/vmware-tanzu/velero/blob/aa7ca1515926fa029e4d2f4d12e712f5e639a9ef/pkg/restore/actions/csi/pvc_action.go#L178

What are your thoughts?

**Comment by kaovilai** [comment]
Sure thing. Works for me.

**Comment by mayankagg9722** [comment]
Great, @kaovilai. I will see if I can pickup and raise the fix for this issue.

**Comment by anshulahuja98** [comment]
https://github.com/vmware-tanzu/velero/pull/8287, @clcondorcet can you go through this discussion.
Is your scenario for non CSI PVs where you are seeing reprovisioning instead of re-attach

This seems to be difference in behaviour as to what is reported in this Issue.

**Comment by mayankagg9722** [comment]
> Sorry for not posting my opinion on this issue earlier, I'm not convinced to make a change in your specified scenario.
> 
> Please check the definitions of backup's `SnapshotVolumes` and restore's `RestorePVs` parameters.
> 
> [velero/pkg/apis/velero/v1/backup_types.go](https://github.com/vmware-tanzu/velero/blob/804d73c4f2349f1ca9bd3d6c751956e1d2021c01/pkg/apis/velero/v1/backup_types.go#L104-L109)
> 
> Lines 104 to 109 in [804d73c](/vmware-tanzu/velero/commit/804d73c4f2349f1ca9bd3d6c751956e1d2021c01)
> 
>  // SnapshotVolumes specifies whether to take snapshots 
>  // of any PV's referenced in the set of objects included 
>  // in the Backup. 
>  // +optional 
>  // +nullable 
>  SnapshotVolumes *bool `json:"snapshotVolumes,omitempty"` 
> [velero/pkg/apis/velero/v1/restore_types.go](https://github.com/vmware-tanzu/velero/blob/804d73c4f2349f1ca9bd3d6c751956e1d2021c01/pkg/apis/velero/v1/restore_types.go#L85-L89)
> 
> Lines 85 to 89 in [804d73c](/vmware-tanzu/velero/commit/804d73c4f2349f1ca9bd3d6c751956e1d2021c01)
> 
>  // RestorePVs specifies whether to restore all included 
>  // PVs from snapshot 
>  // +optional 
>  // +nullable 
>  RestorePVs *bool `json:"restorePVs,omitempty"` 
> To me, the restore's `RestorePVs` should align with the backup's `SnapshotVolumes` setting. According to the definitions, if the backup doesn't enable PV's snapshot-based backup, there is no reason to restore the volume from a snapshot.

Right @blackpiglet. So I am only making a change in removing the volume handle in existing PVC as there was no snapshot found. Velero is restoring is irrespective of my changes.
I dont think keeping PV handle as-is in the cluster is going to help. Correct me if that was an intended behaviour of velero to keep PV name as-is and restore the PVC. (which is not matching with the behaviour while we keep RestorePVs=false)

**Comment by blackpiglet** [comment]
There is a case Velero should restore PVC, and not expect it's dynamically provisioned.

https://github.com/vmware-tanzu/velero/blob/a3cee616dc7201bacb6251ab918fae7dfe994d4e/pkg/restore/restore.go#L1251-L1255

**Comment by mayankagg9722** [comment]
> There is a case Velero should restore PVC, and not expect it's dynamically provisioned.
> 
> [velero/pkg/restore/restore.go](https://github.com/vmware-tanzu/velero/blob/a3cee616dc7201bacb6251ab918fae7dfe994d4e/pkg/restore/restore.go#L1251-L1255)
> 
> Lines 1251 to 1255 in [a3cee61](/vmware-tanzu/velero/commit/a3cee616dc7201bacb6251ab918fae7dfe994d4e)
> 
>  obj, err = ctx.handleSkippedPVHasRetainPolicy(obj, restoreLogger) 
>  if err != nil { 
>  	errs.Add(namespace, err) 
>  	return warnings, errs, itemExists 
>  }

@blackpiglet I have not seen the PV yaml been backed up, so there is no restore of PV YAML here. In the above examples that was already tried. Hence not sure if this code will be called. Correct me if I am missing anything.

https://github.com/vmware-tanzu/velero/blob/a3cee616dc7201bacb6251ab918fae7dfe994d4e/pkg/restore/restore.go#L1204C2-L1204C52

**Comment by blackpiglet** [comment]
@mayankagg9722 
Have you checked the PV YAMLs are not in the backup metadata file?
IMO, PV YAMLs should be backed up.
`SnapshotVolumes` is only related to whether PVs should take snapshots.
PV should be backed up, but restore will make decision whether to create the backed-up PV or dynamically creates new one.

**Comment by mayankagg9722** [comment]
@blackpiglet, I have re-verified the repro. PV YAMLs are getting stored irrespective of if its snapshotted or not.
Also, the code pointer you added is able to handle the PVs with reclaimPolicy delete and able to dynamically provision the PVs in the cluster.

Only corner case which I am seeing where this issue occurred is when during restore _includeClusterResources_ is **False**, PV YAMLs will not be restored as its cluster scoped object and then this issue occurs where PVC is failed during provisioning in the cluster.

**Updating findings: **
1. I have backed up PVs in the **default** namespace. 
2. Restore to **defaultnew** namespace with _includeClusterResources_ as false. **Result**: Failed. PVC provisioning is stuck as PV was already there in other ns in the cluster.
3. and lastly restored to defaultnewcluster namespace with _includeClusterResources_ as true. **Result**: Success


![Image](https://github.com/user-attachments/assets/b00dfe95-5eba-44a3-b30b-72bc5b9f5e6e)

![Image](https://github.com/user-attachments/assets/e1ed0d31-1a40-496c-a4d8-7b7d36065a5b)

Logs:

![Image](https://github.com/user-attachments/assets/64d9e2ef-0f37-4ca6-9675-bd6703e60fa9)

**Code link**:
https://github.com/vmware-tanzu/velero/blob/a3cee616dc7201bacb6251ab918fae7dfe994d4e/pkg/restore/restore.go#L1146

**Comment by sseago** [comment]
@mayankagg9722 Note that for most use cases, you will want `includeClusterResources` to be set to nil (the value you get when it's not set at all in the spec). `false` includes no cluster resources, so as you see, this causes problem with PV restore. `true` includes *all* cluster resources, including those not relevant to your namespaces -- all PVs, all CRDs, etc. The default/auto value (ie. `nil`) only includes relevant cluster-scoped resources. PVs bound to included PVCs, CRDs needed for included CRs, etc.

**Comment by blackpiglet** [comment]
@mayankagg9722 
I agree there would be an error in your specified scenario, but that is a rare case and this is expected by design.

**Comment by mayankagg9722** [comment]
@blackpiglet Do you think setting up volumeName in PVC spec to nil only in case when volume snapshot was not found inside PVC Action has any concerns? **Changes**: https://github.com/vmware-tanzu/velero/pull/8675/files

Or should always latch onto the later codepath in restore flow that you shared while restoring PVs? https://github.com/vmware-tanzu/velero/blob/a3cee616dc7201bacb6251ab918fae7dfe994d4e/pkg/restore/restore.go#L1251-L1255

**Comment by mayankagg9722** [comment]
@blackpiglet should we abandon the PR then as based on the PV deletion policy use case this has already handled while restoring PVs?

**Comment by blackpiglet** [comment]
I do think this PR should not be merged.
